nixos/knot: init

mweinelt opened the pull request:
###### Motivation for this change

Allow for easier usage by creating a module with a secure-by-default systemd unit.


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

mweinelt commented:
I'm still trying to figure out how and where to deploy zonefiles and also where mutable data will be written to. I'm new to Knot DNS and have no experience running it yet.

vcunat commented on nixos/modules/services/networking/knot.nix:
```diff
@@ -0,0 +1,94 @@
+      serviceConfig = {
+        Type = "notify";
+        ExecStart = "${cfg.package}/bin/knotd --config=${knotConfCheck configFile} --socket=${socketFile} ${concatStringsSep " " cfg.extraArgs}";
+        ExecReload = "${cfg.package}/bin/knotc --config=${knotConfCheck configFile} --socket=${socketFile} reload";
```
Why not simplify at least the knotc line by using the wrapped one?
_I'll look at the whole expression a bit more, later; I'm upstream for knot-resolver._

andir replied:
I briefly discussed with @mweinelt on how to do that since he wasn't sure. My recommended way would be changing it as below.

```patch
diff --git a/nixos/modules/services/networking/knot.nix b/nixos/modules/services/networking/knot.nix
index df117071630..9ac7763152c 100644
--- a/nixos/modules/services/networking/knot.nix
+++ b/nixos/modules/services/networking/knot.nix
@@ -13,6 +13,22 @@ let
     ln -s ${configFile} $out
     knotc --config=${configFile} conf-check
   '';
+  cli-wrappers = pkgs.stdenv.mkDerivation {
+    name = "knot-cli-wrappers";
+    buildInputs = [ pkgs.makeWrapper ];
+    buildCommand = ''
+      mkdir -p $out/bin
+      makeWrapper ${cfg.package}/bin/knotc "$out/bin/knotc" \
+        --add-flags "--config=${configFile}" \
+        --add-flags "--socket=${socketFile}"
+      makeWrapper ${cfg.package}/bin/keymgr "$out/bin/keymgr" \
+        --add-flags "--config=${configFile}"
+      for executable in kdig khost kjournalprint knsec3hash knsupdate kzonecheck
+      do
+          ln -s "${cfg.package}/bin/$executable" "$out/bin/$executable"
+      done
+    '';
+  };
 in {
   options = {
     services.knot = {
@@ -55,7 +71,7 @@ in {
       serviceConfig = {
         Type = "notify";
         ExecStart = "${cfg.package}/bin/knotd --config=${knotConfCheck configFile} --socket=${socketFile} ${concatStringsSep " " cfg.extraArgs}";
-        ExecReload = "${cfg.package}/bin/knotc --config=${knotConfCheck configFile} --socket=${socketFile} reload";
+        ExecReload = "${cli-wrappers}/bin/knotc reload";
         CapabilityBoundingSet="CAP_NET_BIND_SERVICE CAP_SETPCAP";
         AmbientCapabilities="CAP_NET_BIND_SERVICE CAP_SETPCAP";
         NoNewPrivileges=true;
@@ -69,24 +85,7 @@ in {
       };
     };
 
-   environment.systemPackages = let
-      cli-wrappers = pkgs.stdenv.mkDerivation {
-        name = "knot-cli-wrappers";
-        buildInputs = [ pkgs.makeWrapper ];
-        buildCommand = ''
-          mkdir -p $out/bin
-          makeWrapper ${cfg.package}/bin/knotc "$out/bin/knotc" \
-            --add-flags "--config=${configFile}" \
-            --add-flags "--socket=${socketFile}"
-          makeWrapper ${cfg.package}/bin/keymgr "$out/bin/keymgr" \
-            --add-flags "--config=${configFile}"
-          for executable in kdig khost kjournalprint knsec3hash knsupdate kzonecheck
-          do
-              ln -s "${cfg.package}/bin/$executable" "$out/bin/$executable"
-          done
-        '';
-      };
-    in [ cli-wrappers ];
+   environment.systemPackages = [ cli-wrappers ];
 
   };
 ```

mweinelt replied:
Applied, thanks!

andir commented on nixos/modules/services/networking/knot.nix:
```diff
@@ -0,0 +1,94 @@
+
+      extraConfig = mkOption {
+        type = types.lines;
+        default = false;
```
The default value of `false` causes an error when enabling the services without any configuration:

```
error: The option value `services.knot.extraConfig' in `/home/andi/dev/nixos/nixpkgs-review/nixos/modules/services/networking/knot.nix' is not of type `strings concatenated with "\n"'.
```

It should probably be "" or `types.nullOr types.lines` with some additional logic to handle the empty config file (if that is even valid).

mweinelt replied:
Knot seems to be fine with an empty config, so if that is sufficient to enable the service I'd stick with `""`.

andir reviewed (commented):
Overall it looks like a good first draft. Thanks for working on it! :+1: 

When running the daemon with an empty configuration I received a bunch of warnings which we might want to investigate:

```
server# [    8.219413] knotd[792]: 2019-03-06T14:06:29 info: Knot DNS 2.7.6 starting
server# [    8.221783] knotd[792]: 2019-03-06T14:06:29 info: loading 0 zones
server# [    8.233685] knotd[792]: warning: cannot open persistent timer DB '/nix/store/0sv2dhwmfr3q7y67n2q0jfqipmfmjm5r-knot-dns-2.7.6/var/lib/knot/timers' (not exists)
server# [    8.238169] knotd[792]: 2019-03-06T14:06:29 warning: cannot open persistent timer DB '/nix/store/0sv2dhwmfr3q7y67n2q0jfqipmfmjm5r-knot-dns-2.7.6/var/lib/knot/timers' (not exists)
server# [    8.241953] knotd[792]: 2019-03-06T14:06:29 warning: no zones loaded
server# [    8.244193] knotd[792]: 2019-03-06T14:06:29 info: starting server
server# [    8.246495] knotd[792]: 2019-03-06T14:06:29 info: server started in the foreground, PID 792
server# [    8.249342] knotd[792]: 2019-03-06T14:06:29 info: control, binding to '/run/knot/knot.sock'
server# [    8.252291] knotd[792]: warning: no zones loaded
```

mweinelt commented:
Thanks, I'm looking into the feedback tonight.

vcunat commented:
> /nix/store/HASH-knot-dns-VERSION/var

It's possible we'll want to modify the `knot-dns` package expression.  So far I wrote it because of libs (for `knot-resolver`) and the `kdig` tool, so I haven't paid much attention to other parts.

mweinelt commented:
> warning: cannot open persistent timer DB '/nix/store/0sv2dhwmfr3q7y67n2q0jfqipmfmjm5r-knot-dns-2.7.6/var/lib/knot/timers' (not exists)

The timer implementation changed between between 2.7.6 and 2.8.0 and I guess we should probably bump to 2.8.0 before investigating. https://gitlab.labs.nic.cz/knot/knot-dns/commit/844209ef4fb68b582648871c44dd82fb3ce2e3b7

vcunat commented:
There's a catch for knot-resolver, so I opened https://github.com/NixOS/nixpkgs/pull/56974 to at least serve as base for developing the nixos service.

mweinelt commented:
Yup, the timer warnings are gone on 2.8.0.

```
Mar 06 19:30:41 juno systemd[1]: Starting Authoritative-only DNS server from .cz domain registry...
Mar 06 19:30:41 juno knotd[11785]: 2019-03-06T19:30:41 info: Knot DNS 2.8.0 starting
Mar 06 19:30:41 juno knotd[11785]: 2019-03-06T19:30:41 info: loading 0 zones
Mar 06 19:30:41 juno knotd[11785]: 2019-03-06T19:30:41 warning: no zones loaded
Mar 06 19:30:41 juno knotd[11785]: 2019-03-06T19:30:41 info: starting server
Mar 06 19:30:41 juno knotd[11785]: 2019-03-06T19:30:41 info: server started in the foreground, PID 11785
Mar 06 19:30:41 juno knotd[11785]: 2019-03-06T19:30:41 info: control, binding to '/run/knot/knot.sock'
Mar 06 19:30:41 juno knotd[11785]: warning: no zones loaded
Mar 06 19:30:41 juno systemd[1]: Started Authoritative-only DNS server from .cz domain registry.
```

mweinelt commented on nixos/modules/services/networking/knot.nix:
```diff
@@ -0,0 +1,94 @@
+        Type = "notify";
+        ExecStart = "${cfg.package}/bin/knotd --config=${knotConfCheck configFile} --socket=${socketFile} ${concatStringsSep " " cfg.extraArgs}";
+        ExecReload = "${cfg.package}/bin/knotc --config=${knotConfCheck configFile} --socket=${socketFile} reload";
+        CapabilityBoundingSet="CAP_NET_BIND_SERVICE CAP_SETPCAP";
```
Indeed, thanks for noticing. Cleaned up.

vcunat commented:
@andir: what about using this?
```nix
{ # Knot DNS uses a restricted version of YAML.
  nix2yaml = let
    quoteString = s: ''"${toString s}"'';
    n2y = indent: val:
      if isAttrs val then concatStringsSep "\n${indent}"
        (mapAttrsToList
          # This is a bit wacky - set directly under a set would start on bad indent,
          # so we start those on a new line, but not other types of attribute values.
          (aname: aval: "${aname}:${if isAttrs aval then "\n${indent}  " else " "}"
            + nix2yaml_g (indent + "  ") aval)
          val
        )
        + "\n"
        else
      if isList val && stringLength indent < 4 then concatMapStrings
        (elem: "\n${indent}- " + nix2yaml_g (indent + "  ") elem)
        val
        else
      if isList val /* and long indent */ then
        "[ " + concatMapStringsSep ", " quoteString val + " ]" else
      quoteString val;
    in n2y "";
}
```

Example output:
```yaml
server:
  listen: [ "0.0.0.0@53", ":@53" ]

zone: 
  - domain: "deleg.example.com."
    file: "/nix/store/2vn5pj4hsbsjbwr9is3syb01r7zsbcqn-deleg.example.com."
    storage: "/var/lib/knot"

  - domain: "example.com."
    file: "/nix/store/f9cxcqcw3kxk0pb8nv6a8dahxbzdfs44-example.com."
    storage: "/var/lib/knot"
```

andir commented:
@vcunat that looks great and (almost) works for my test that I hacked together :+1:. I had to convert paths to string before passing it to the function but that is a trivial thing to add.

vcunat commented:
I thought `toString` worked fine on paths, except it would try to clone them into the `/nix/store` if written as such... which is what you meant?  (But that feels like a natural semantics in nix to me.)

vcunat commented:
Ah, it's the other way around... I break the usual semantics by applying `toString` to paths as well.

andir commented:
I was using the following during my tests:

```nix
let
      zones."example.com." = ''
        @ SOA ns.example.com noc.example.com 666 7200 3600 1209600 3600
        ipv4 A 1.2.3.4
        ipv6 AAAA abcd::eeff
        deleg NS ns.example.com
        ns A 192.168.0.1
        ns AAAA dead:beef::1
      '';
      zones."deleg.example.com." = ''
        @ SOA ns.example.com noc.example.com 666 7200 3600 1209600 3600
        @ A 9.8.7.6
        @ AAAA fedc::bbaa
      '';
      #zones."." = ''
      #  @ SOA ns.example.com noc.example.com 666 7200 3600 1209600 3600
      #  root A 1.8.7.4
      #  root AAAA acbd::4
      #'';
      nix2yaml = with lib; let
        quoteString = s: ''"${toString s}"'';
        n2y = indent: val:
          if isAttrs val then concatStringsSep "\n${indent}"
            (mapAttrsToList
              # This is a bit wacky - set directly under a set would start on bad indent,
              # so we start those on a new line, but not other types of attribute values.
              (aname: aval: "${aname}:${if isAttrs aval then "\n${indent}  " else " "}"
                + n2y (indent + "  ") aval)
              val
            )
            + "\n"
            else
          if isList val && stringLength indent < 4 then concatMapStrings
            (elem: "\n${indent}- " + n2y (indent + "  ") elem)
            val
            else
          if isList val /* and long indent */ then
            "[ " + concatMapStringsSep ", " quoteString val + " ]" else
          quoteString val;
      in n2y "";
    in {
      networking.nameservers = lib.mkForce [ "192.168.0.1" ];
      networking.interfaces.eth1.ipv4.addresses = [
        { address = "192.168.0.1"; prefixLength = 24; }
      ];
      networking.interfaces.eth1.ipv6.addresses = [
        { address = "dead:beef::1"; prefixLength = 64; }
      ];

      services.knot = {
        enable = true;
        extraConfig = nix2yaml {
          server.listen = [ "0.0.0.0@53" "::@53" ];
          zone = lib.mapAttrsToList (n: v: { domain = n; storage = "/var/lib/knot"; file = toString (pkgs.writeText n v); }) zones;
        };
      };
```

I added the extra `toString`to the `pkgs.writeText` line. Since `lib.isAttrs` detects the result of `writeText` as an attribute set (rightfully?) and tried to access it's attributes which are functions.

vcunat commented:
Right, there were some subtle issues, commented in the code.
```nix
rec { # Knot DNS uses a restricted version of YAML.
  nix2yaml = let
    # We don't want paths like ./my-zone.txt be converted to plain strings.
    quoteString = s: ''"${if builtins.typeOf s == "path" then s else toString s}"'';
    # We don't want to walk the insides of derivation attributes.
    doRecurse = val: isAttrs val && !isDerivation val;
    n2y = indent: val:
      if doRecurse val then concatStringsSep "\n${indent}"
        (mapAttrsToList
          # This is a bit wacky - set directly under a set would start on bad indent,
          # so we start those on a new line, but not other types of attribute values.
          (aname: aval: "${aname}:${if doRecurse aval then "\n${indent}  " else " "}"
            + n2y (indent + "  ") aval)
          val
        )
        + "\n"
        else
      if isList val && stringLength indent < 4 then concatMapStrings
        (elem: "\n${indent}- " + n2y (indent + "  ") elem)
        val
        else
      if isList val /* and long indent */ then
        "[ " + concatMapStringsSep ", " quoteString val + " ]" else
      quoteString val;
    in n2y "";


       extraConfig = {
          server.listen = [ "0.0.0.0@53" "::@53" ];
          zone = mapAttrsToList
            (n: v: {
              domain = n;
              storage = "/var/lib/knot";
              file = if isString v then pkgs.writeText n v else v;
            })
            zones;
        };
       # which works with these options that came into my mind:
       zones.foo = "inline string";
       zones.bar = ./some-file;
       zones.baz = some derivation; # e.g. writeText, fetch* or whatever
}
```

vcunat commented:
I guess we could make a similar "zone.*" processing built into the conversion into YAML (as a pre-processor map on nix level), so that people have an option to configure it in a way that's more natural in nix.

andir commented:
@vcunat how do you feel about merging the current implementation once it gets a test and then only afterwards investing in a more "complete" nix integration?

I think it would make sense to experiment with the different options before putting each through a nixpkgs PR. My personal DNS servers will move to this as soon as we merge this and I get the time to roll it out.

vcunat commented:
Yes, I'm in favor of faster iteration.

vcunat commented:
The changes might be (mostly) backward compatible even, if we only support the basic config options now &ndash; those that don't affect the configuration file.  But anyway, I don't expect people will massively start to use the new service, so 100% compatibility doesn't seem too important.

vcunat commented:
I reviewed the systemd unit config, comparing it to upstream unit.  Minor changes:
- `NoNewPrivileges = true;  PrivateDevices = true;` these seem OK
- `RestrictAddressFamilies = "AF_UNIX AF_INET AF_INET6";`
I'm perhaps missing the point there.  And according to docs it's not much use without `SystemCallArchitectures=native`, because of the former _not_ restricting 32-bit x86 syscalls.
- `StartLimitInterval = "0";`  Can you explain why?  Are restart bursts required for something?  BTW, the option [has been renamed](https://github.com/systemd/systemd/commit/f0367da7d1a61).

Now, the change I consider big: `DynamicUser = "yes";`.  Well, I can't claim to have much experience using this option or running authoritative DNS, but it seems quite a non-standard setup at least.  Can you explain why not use a standard "persistent" user instead?  I couldn't quickly find in docs in what way the state directory is being kept between service restarts, etc.

andir commented:
> Now, the change I consider big: DynamicUser = "yes";. Well, I can't claim to have much experience using this option or running authoritative DNS, but it seems quite a non-standard setup at least. Can you explain why not use a standard "persistent" user instead? I couldn't quickly find in docs in what way the state directory is being kept between service restarts, etc.

The idea is to avoid issues with both exhausting our UID/GID preallocation space (that is rather limited) and to avoid any kind of UID/GID reuse attacks. Systemd will take care of making the StateDirectory (usually /var/lib/<name>) writable for the dynamic user. We had a few discussions on IRC (#nixos{,-dev,-chat,-security,…}) where  the conclusion was that dynamic users + state dir might be desireable for new services. There isn't anything documented about that idea just yet.

On that note we might want to set `StateDirectoryMode=` to something that isn't 0755 to protect the private keys - if know doesn't do that already.

vcunat commented:
So the state directory will still be persistent and kept across service/system restarts, right?  (Changing the ownership in case systemd starts it with a different UID/GID, etc.)

andir commented:
Yes


On Mon, Mar 11, 2019 at 9:22 AM Vladimír Čunát <notifications@github.com>
wrote:

> So the state directory will still be persistent and kept across
> service/system restarts, right? (Changing the ownership in case systemd
> starts it with a different UID/GID, etc.)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/56922#issuecomment-471463082>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAm_dHiLYFK4A_TQbd4Qa9CAFEheoAj0ks5vViBmgaJpZM4bfrm9>
> .
>

vcunat commented:
OK, sounds good (with my limited knowledge).  Dir permissions sound OK as well, according to `man systemd.exec`, but we'd better also check it "in real life" :-)
> If DynamicUser= is used in conjunction with StateDirectory=, CacheDirectory= and LogsDirectory=
           is slightly altered: the directories are created below /var/lib/private, /var/cache/private and
           /var/log/private, respectively, which are host directories made **inaccessible to unprivileged
           users**, which ensures that access to these directories cannot be gained through dynamic user ID
           recycling.

mweinelt commented:
Thanks for your reviews. I made the following changes:

-  Added `SystemCallArchitectures=native` to prevent circumvention of `RestrictAddressFamilies` via syscalls against non-native ABIs
- Dropped `StartLimitInterval`, it was a leftover from copying and pasting an example systemd unit description
- Restricted `StateDirectoryMode` from `0755` to `0700`

I'm looking into tests right now. Also I'm still wondering what (the package, the module, something else?) is responsible for getting man pages into the system environment.

mweinelt commented:
Regarding the state directories permission:

```shell
[root@juno:~]$ ls -lah /var/lib/knot
lrwxrwxrwx 1 root root 12 Mar  3 02:58 /var/lib/knot -> private/knot
```

```shell
[root@juno:~]$ ls -lah /var/lib/private/knot/
total 8.0K
drwx------ 2 knot knot 4.0K Mar  3 02:58 .
drwx------ 4 root root 4.0K Mar  3 02:58 ..
```

```shell
[hexa@juno:~]$ ls -lah /var/lib/private/knot
ls: cannot access '/var/lib/private/knot': Permission denied
```

vcunat commented on nixos/modules/services/networking/knot.nix:
```diff
@@ -0,0 +1,93 @@
+      for executable in kdig khost kjournalprint knsec3hash knsupdate kzonecheck
+      do
+        ln -s "${cfg.package}/bin/$executable" "$out/bin/$executable"
+      done
```
Man pages: perhaps I'd simply add this here:
```bash
mkdir -p "$out/share"
ln -s '${knot-dns.bin}/share/man' "$out/share/"
```
The closure remains the same anyway.

vcunat replied:
Actually, what we could additionally build is the html docs, i.e. offline version of https://www.knot-dns.cz/documentation/  I can modify the `knot-dns` expression to produce those, if you'd find them useful.

mweinelt replied:
I've actually always viewed more extensive documentation on project websites. Especially since servers usually run headless and are configured from machines that have access to a browser.

vcunat replied:
Right... it would be more like installing the docs (without the server) on the dev machine while having just the binaries (without docs) on the server.  I suppose people are used to online docs nowadays.

mweinelt replied:
I added a link to the version-specific HTML documentation to the systemd units `Documentation=` key.

vcunat commented on nixos/modules/services/networking/knot.nix:
```diff
@@ -0,0 +1,93 @@
+  cfg = config.services.knot;
+
+  configFile = pkgs.writeText "knot.conf" cfg.extraConfig;
+  socketFile = "/run/knot/knot.sock";
```
Nitpick: thanks to `DynamicUser = "yes";` there won't be anything else in `/run/knot` AFAIK, so we might just call it `"/run/knot.sock"`, though the less-standard location might not be worth this simplification.

mweinelt commented:
The last commit adds tests, I adapted the test set already used by nsd.

- server on 192.168.0.1/24 and fd00::1/64
  - knotd binds against IPv4 and IPv6 wildcards
- client on 192.168.0.2/24 and fd00::2/64
  - interrogates knotd on IPv4 and IPv6 using khost
    - tests for missing A and AAAA
    - tests for SOA, A and AAA
    - tests NS delegation by delegating `sub.example.com` to itself

mweinelt commented:
The last commit links the man pages into the system environment, thanks to @vcunat.

mweinelt committed:
nixos/knot: init

mweinelt commented:
I just tried enabling DNSSEC signing and the initialization failed because my zones currently reside in the nix store and Knot seems to want to write data in the same directory. I will probably need to symlink the zones somewhere below the `StateDirectory` instead.

```
knotd[20282]: info: loading 1 zones
knotd[20282]: info: [_acme.example.com.] zone will be loaded
knotd[20282]: info: starting server
knotd[20282]: info: [_acme.example.com.] zone file parsed, serial 1552483525
knotd[20282]: error: [_acme.example.com.] DNSSEC, failed to initialize (Read-only file system)
knotd[20282]: error: [_acme.example.com.] zone event 'load' failed (Read-only file system)
knotd[20282]: info: server started in the foreground, PID 20282
```

vcunat commented:
Hmm, perhaps [this approach](https://www.knot-dns.cz/docs/2.8/singlehtml/index.html#example-3)?  For a master server, that is.  For slaves the preceding example would probably be suitable, but we haven't really discussed slaves here yet.

mweinelt commented:
No, unfortunately that does not seem to change the behaviour.

mweinelt commented:
The tests now enable `dnssec-signing` with the workarounds mentioned in https://github.com/NixOS/nixpkgs/pull/56922#issuecomment-472484268. it is now checking for the existence of a DNSKEY record on the zone apex and a RRSIG record on the `www` resource record.

mweinelt commented:
I just pushed the updated tests with the master/slave setup. Is there anything else?

vcunat reviewed (approved):
I can't see anything else ATM.  Let's reiterate if needed.

vcunat commented:
Closely related: https://discourse.nixos.org/t/nix-dns-a-nix-dsl-for-dns-zone-files/2466

nixos-discourse commented:
This pull request has been mentioned on **Nix community**. There might be relevant details there:

https://discourse.nixos.org/t/generating-yaml-in-a-nix-module/2519/5

nixos-discourse commented:
This pull request has been mentioned on **Nix community**. There might be relevant details there:

https://discourse.nixos.org/t/generating-yaml-in-a-nix-module/2519/5
